File: repair.go
package main

import (
	"bufio"
	"encoding/binary"
	"fmt"
	"io"
	"io/ioutil"
	"math"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"time"
)

// cutWavHeader reads the first 44 bytes of the WAV file, which contains the header
func cutWavHeader(inputPath string) ([]byte, error) {
	file, err := os.Open(inputPath)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	header := make([]byte, 44)
	_, err = file.Read(header)
	if err != nil {
		return nil, err
	}

	return header, nil
}

// loadEncryptedWav reads the WAV file data starting from the given offset to -334 bytes from the end
func loadEncryptedWav(inputPath string, offset int) ([]byte, error) {
	file, err := os.Open(inputPath)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	stat, err := file.Stat()
	if err != nil {
		return nil, err
	}

	fileSize := stat.Size()
	buffer := make([]byte, fileSize-334-int64(offset))

	_, err = file.Seek(int64(offset), io.SeekStart)
	if err != nil {
		return nil, err
	}

	_, err = file.Read(buffer)
	if err != nil && err != io.EOF {
		return nil, err
	}

	return buffer, nil
}

// mergeWavFiles combines the header and data to form the complete WAV file
func mergeWavFiles(referenceHeader, encryptedData []byte) []byte {
	return append(referenceHeader, encryptedData...)
}

// saveRepairedWav writes the repaired WAV data to a new file
func saveRepairedWav(inputPath string, repairedData []byte, riffChunkSize, dataChunkSize uint32, outputFolder string) error {
	err := os.MkdirAll(outputFolder, 0755)
	if err != nil {
		return err
	}

	inputFilename := strings.TrimSuffix(filepath.Base(inputPath), filepath.Ext(inputPath))
	if filepath.Ext(inputFilename) == ".wav" {
		inputFilename = strings.TrimSuffix(inputFilename, filepath.Ext(inputFilename))
	}
	outputFullPath := filepath.Join(outputFolder, inputFilename+".wav")

	header := repairedData[:44]
	binary.LittleEndian.PutUint32(header[4:8], riffChunkSize)
	binary.LittleEndian.PutUint32(header[40:44], dataChunkSize)

	file, err := os.Create(outputFullPath)
	if err != nil {
		return err
	}
	defer file.Close()

	writer := bufio.NewWriter(file)
	_, err = writer.Write(header)
	if err != nil {
		return err
	}

	_, err = writer.Write(repairedData[44:])
	if err != nil {
		return err
	}

	err = writer.Flush()
	if err != nil {
		return err
	}

	fmt.Printf("File saved to %s\n", outputFullPath)
	return nil
}

// calculateFirstCompleteFramePosition calculates the position of the first complete frame in the WAV file
func calculateFirstCompleteFramePosition(bitsPerSample, numChannels uint16) int {
	frameSize := (int(bitsPerSample) / 8) * int(numChannels)
	firstCompleteFrameNumber := int(math.Ceil(float64(153605-44) / float64(frameSize)))
	return 44 + (firstCompleteFrameNumber * frameSize)
}

// isDoubleExtensionWav checks if the file has a double extension ending in .wav
func isDoubleExtensionWav(fileName string) bool {
	nameWithoutExt := strings.TrimSuffix(fileName, filepath.Ext(fileName))
	secondExt := filepath.Ext(nameWithoutExt)
	return strings.ToLower(secondExt) == ".wav"
}

// processFile processes a single WAV file, repairing it and saving the result
func processFile(wg *sync.WaitGroup, referenceHeader []byte, position int, encryptedWavPath, outputFolder string) {
	defer wg.Done()

	encryptedData, err := loadEncryptedWav(encryptedWavPath, position)
	if err != nil {
		fmt.Println("Error loading encrypted WAV data:", err)
		return
	}

	fileSize := len(encryptedData) + 44
	riffChunkSize := uint32(fileSize - 8)
	dataChunkSize := uint32(fileSize - 44)

	repairedWavData := mergeWavFiles(referenceHeader, encryptedData)

	err = saveRepairedWav(encryptedWavPath, repairedWavData, riffChunkSize, dataChunkSize, outputFolder)
	if err != nil {
		fmt.Println("Error saving repaired WAV file:", err)
	}
}

func main() {
	cwd, err := os.Getwd()
	if err != nil {
		fmt.Println("Error getting current working directory:", err)
		return
	}
	fmt.Println("Current working directory:", cwd)

	var referenceWavPath string
	fmt.Print("Enter the path to the reference WAV file: ")
	fmt.Scanln(&referenceWavPath)

	startTime := time.Now()

	referenceHeader, err := cutWavHeader(referenceWavPath)
	if err != nil {
		fmt.Println("Error reading reference WAV header:", err)
		return
	}

	bitsPerSample := binary.LittleEndian.Uint16(referenceHeader[34:36])
	numChannels := binary.LittleEndian.Uint16(referenceHeader[22:24])
	position := calculateFirstCompleteFramePosition(bitsPerSample, numChannels)
	fmt.Println("Position of the first complete frame:", position)

	var encryptedWavFolder string
	fmt.Print("Enter the path to the folder containing encrypted WAV files: ")
	fmt.Scanln(&encryptedWavFolder)

	corruptedFilesFolder := filepath.Dir(encryptedWavFolder)
	outputFolder := filepath.Join(corruptedFilesFolder, "Repaired")

	files, err := ioutil.ReadDir(encryptedWavFolder)
	if err != nil {
		fmt.Println("Error reading encrypted WAV folder:", err)
		return
	}

	var wg sync.WaitGroup
	for _, file := range files {
		fmt.Println("Checking file:", file.Name())
		if isDoubleExtensionWav(file.Name()) {
			wg.Add(1)
			go processFile(&wg, referenceHeader, position, filepath.Join(encryptedWavFolder, file.Name()), outputFolder)
		}
	}
	wg.Wait()

	endTime := time.Now()
	fmt.Printf("Execution time: %.2f seconds\n", endTime.Sub(startTime).Seconds())
}

File: repair.pl
use strict;
use warnings;
use File::Path qw(make_path);
use File::Copy;
use File::Basename;

sub cut_wav_header {
    my ($input_path) = @_;
    open(my $fh, '<:raw', $input_path) or die "Cannot open file $input_path: $!";
    read($fh, my $header, 44);
    close($fh);
    return $header;
}

sub load_encrypted_wav {
    my ($input_path, $offset) = @_;
    open(my $fh, '<:raw', $input_path) or die "Cannot open file $input_path: $!";
    seek($fh, $offset, 0);
    read($fh, my $encrypted_data, -s $fh - 334);  # Read until end of file minus 334 bytes
    close($fh);
    return $encrypted_data;
}

sub merge_wav_files {
    my ($reference_header, $encrypted_data) = @_;
    return $reference_header . $encrypted_data;
}

sub save_repaired_wav {
    my ($input_path, $repaired_data, $riff_chunk_size, $data_chunk_size, $output_folder) = @_;

    make_path($output_folder);

    my ($input_filename, $input_dirs, $suffix) = fileparse($input_path, qr/\.[^.]*/);
    my $output_full_path = "$output_folder/$input_filename.wav";

    my $header = substr($repaired_data, 0, 44);
    substr($header, 4, 4) = pack('V', $riff_chunk_size);
    substr($header, 40, 4) = pack('V', $data_chunk_size);

    open(my $fh, '>:raw', $output_full_path) or die "Cannot create file $output_full_path: $!";
    print $fh $header . substr($repaired_data, 44);
    close($fh);

    print "File saved to $output_full_path\n";
}

sub calculate_first_complete_frame_position {
    my ($bits_per_sample, $num_channels) = @_;
    my $frame_size = int($bits_per_sample / 8) * $num_channels;
    my $first_complete_frame_number = int((153605 - 44) / $frame_size) + 1;
    return 44 + ($first_complete_frame_number * $frame_size);
}

sub is_double_extension_wav {
    my ($file_name) = @_;
    my ($name_without_ext, $first_ext) = fileparse($file_name, qr/\.[^.]*/);
    my ($second_name, $second_ext) = fileparse($name_without_ext, qr/\.[^.]*/);
    return lc($second_ext) eq '.wav';
}

sub main {
    print "Current working directory: " . `cd`;
    print "Enter the path to the reference WAV file: ";
    my $reference_wav_path = <STDIN>;
    chomp($reference_wav_path);

    my $start_time = time();

    my $reference_header = cut_wav_header($reference_wav_path);
    my $bits_per_sample = unpack('v', substr($reference_header, 34, 2));
    my $num_channels = unpack('v', substr($reference_header, 22, 2));
    my $position = calculate_first_complete_frame_position($bits_per_sample, $num_channels);
    print "Position of the first complete frame: $position\n";

    print "Enter the path to the folder containing encrypted WAV files: ";
    my $encrypted_wav_folder = <STDIN>;
    chomp($encrypted_wav_folder);

    my $corrupted_files_folder = dirname($encrypted_wav_folder);
    my $output_folder = "$corrupted_files_folder/Repaired";

    opendir(my $dh, $encrypted_wav_folder) or die "Cannot open directory $encrypted_wav_folder: $!";
    while (my $file = readdir($dh)) {
        next if $file =~ /^\./;
        print "Checking file: $file\n";
        if (is_double_extension_wav($file)) {
            my $encrypted_wav_path = "$encrypted_wav_folder/$file";
            my $offset = $position;
            my $encrypted_data = load_encrypted_wav($encrypted_wav_path, $offset);

            my $file_size = length($encrypted_data) + 44;
            my $riff_chunk_size = $file_size - 8;
            my $data_chunk_size = $file_size - 44;

            my $repaired_wav_data = merge_wav_files($reference_header, $encrypted_data);

            save_repaired_wav($encrypted_wav_path, $repaired_wav_data, $riff_chunk_size, $data_chunk_size, $output_folder);
        }
    }
    closedir($dh);

    my $end_time = time();
    my $execution_time = $end_time - $start_time;
    printf "Execution time: %.2f seconds\n", $execution_time;
}

main();
